x/evm/keeper/keeper.go

```diff
@@ -213,166 +174,295 @@ func (k Keeper) GetTxIndexTransient() uint64 {
+			return err
+		}
+	default:
+		// not changed
+	}
+	return nil
 }
 
-// ResetAccount removes the code, storage state, but keep all the native tokens stored
-// with the given address.
-func (k Keeper) ResetAccount(addr common.Address) {
-	k.DeleteCode(addr)
-	k.DeleteAccountStorage(addr)
+// SetAccount updates nonce/balance/codeHash together.
+func (k *Keeper) SetAccount(ctx sdk.Context, addr common.Address, account statedb.Account) error {
+	// update account
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	acct := k.accountKeeper.GetAccount(ctx, cosmosAddr)
+	if acct == nil {
+		acct = k.accountKeeper.NewAccountWithAddress(ctx, cosmosAddr)
+	}
+	ethAcct, ok := acct.(*ethermint.EthAccount)
+	if !ok {
+		return errors.New("not EthAccount")
+	}
+	if err := ethAcct.SetSequence(account.Nonce); err != nil {
+		return err
+	}
+	ethAcct.CodeHash = common.BytesToHash(account.CodeHash).Hex()
+	k.accountKeeper.SetAccount(ctx, ethAcct)
+
+	return k.SetBalance(ctx, addr, account.Balance)
 }
 
-// SetHooks sets the hooks for the EVM module
-func (k *Keeper) SetHooks(eh types.EvmHooks) *Keeper {
-	if k.hooks != nil {
-		panic("cannot set evm hooks twice")
+func (k *Keeper) SetState(ctx sdk.Context, addr common.Address, key, value common.Hash) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AddressStoragePrefix(addr))
+	if len(value) == 0 {
+		store.Delete(key.Bytes())
+	} else {
+		store.Set(key.Bytes(), value.Bytes())
 	}
+}
 
-	k.hooks = eh
-	return k
+func (k *Keeper) SetCode(ctx sdk.Context, codeHash []byte, code []byte) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixCode)
+
+	// store or delete code
+	if len(code) == 0 {
+		store.Delete(codeHash)
+	} else {
+		store.Set(codeHash, code)
+	}
 }
 
-// PostTxProcessing delegate the call to the hooks. If no hook has been registered, this function returns with a `nil` error
-func (k *Keeper) PostTxProcessing(from common.Address, to *common.Address, receipt *ethtypes.Receipt) error {
-	if k.hooks == nil {
+// PurgeAccount handles contract's suicide call:
+// - clear balance
+// - remove code
+// - remove states
+func (k *Keeper) PurgeAccount(ctx sdk.Context, addr common.Address) error {
+	acct, err := k.GetAccount(ctx, addr)
+	if err != nil {
+		return err
+	}
+
+	if acct == nil {
 		return nil
 	}
-	return k.hooks.PostTxProcessing(k.Ctx(), from, to, receipt)
-}
 
-// Tracer return a default vm.Tracer based on current keeper state
-func (k Keeper) Tracer(msg core.Message, ethCfg *params.ChainConfig) vm.Tracer {
-	return types.NewTracer(k.tracer, msg, ethCfg, k.Ctx().BlockHeight())
+	// clear balance
+	if acct.Balance.Sign() != 0 {
+		if err := k.SetBalance(ctx, addr, new(big.Int)); err != nil {
+			return err
+		}
+	}
+
+	// remove code
+	if !bytes.Equal(acct.CodeHash, emptyCodeHash) {
+		k.SetCode(ctx, acct.CodeHash, []byte{})
+	}
+
+	// clear storage
+	k.ForEachStorage(ctx, addr, func(key, value common.Hash) bool {
+		k.SetState(ctx, addr, key, common.Hash{})
```

Comment:
delete state?

Reply:
Both `SetCode` and `SetState` has the logic which do deletion if `len(value) == 0`.